File: src/wallet.py
"""
Keys, addresses, and signatures for DeCoin.

DeCoin shipped without any of this: transactions carried a `signature` field that
was never set and never checked, and "addresses" were arbitrary strings, so any
address was spendable by anyone. This module is the missing cryptography — real
ECDSA over secp256k1, the same curve Bitcoin uses.

An address is derived from a public key the Bitcoin way — RIPEMD160(SHA256(pub)) —
so it commits to the key that controls it. Signing covers a transaction's
canonical content, and verification checks two things together: that the
signature is valid for the stated public key, and that the public key hashes to
the sender address. Only the holder of the private key can produce that pair.
"""
import hashlib
import json
from typing import List, Optional

from ecdsa import SigningKey, VerifyingKey, SECP256k1, BadSignatureError

# Verification runs on every node for every signature in every block, so it is
# the hot path at scale. Pure-Python ecdsa verifies ~1,000/s; libsecp256k1 (the
# C library Bitcoin and Ethereum use) does tens of thousands per core. Prefer it
# when installed and fall back to pure Python so the project still clones and
# runs with no compiled dependency. SIGNING stays on `ecdsa` either way, so the
# signatures a wallet produces are byte-identical across backends and across
# languages (the browser wallet must reproduce them exactly).
try:
    from coincurve import PublicKey as _CCPublicKey
    from coincurve.ecdsa import (
        deserialize_compact as _cc_deserialize_compact,
        cdata_to_der as _cc_cdata_to_der,
        signature_normalize as _cc_signature_normalize,
    )
    _HAVE_COINCURVE = True
except ImportError:  # pragma: no cover - exercised only where coincurve is absent
    _HAVE_COINCURVE = False

ADDRESS_PREFIX = "DEC"
MULTISIG_PREFIX = "DECMS"


def _sha256d(message: bytes) -> bytes:
    return hashlib.sha256(message).digest()


def address_from_public_key(public_key_bytes: bytes) -> str:
    """RIPEMD160(SHA256(pubkey)), hex, with the DEC prefix. The address is a
    commitment to the public key, which is why a signature can prove ownership
    of the address without the address ever revealing the key."""
    sha = hashlib.sha256(public_key_bytes).digest()
    ripe = hashlib.new("ripemd160", sha).hexdigest()
    return ADDRESS_PREFIX + ripe


def address_from_public_key_hex(public_key_hex: str) -> str:
    return address_from_public_key(bytes.fromhex(public_key_hex))


def multisig_address(signer_addresses: List[str], required: int) -> str:
    """An m-of-n address commits to WHO can sign and HOW MANY are needed, so the
    address cannot be spent by a different set of signers or a lower threshold.
    Order-independent (signers are sorted) so every party derives the same one."""
    payload = json.dumps(
        {"signers": sorted(signer_addresses), "required": int(required)},
        sort_keys=True,
    ).encode()
    ripe = hashlib.new("ripemd160", hashlib.sha256(payload).digest()).hexdigest()
    return MULTISIG_PREFIX + ripe


def verify_signature(public_key_hex: str, signature_hex: str, message: bytes) -> bool:
    """True only if `signature_hex` is a valid secp256k1 signature over `message`
    for `public_key_hex`. Any malformed input is a verification failure, never an
    exception the caller has to guard.

    Both backends verify the same signatures: a 64-byte compact (r||s) signature
    over SHA-256 of the message, for a 64-byte raw public key. The signature may
    carry a high S value (the signing rule does not normalise it); a high-S
    signature is mathematically valid, so both paths accept it — libsecp256k1
    after an explicit low-S normalisation, which changes acceptance, not validity."""
    try:
        pub = bytes.fromhex(public_key_hex)
        sig = bytes.fromhex(signature_hex)
    except (ValueError, TypeError):
        return False
    if len(pub) != 64 or len(sig) != 64:
        return False

    if _HAVE_COINCURVE:
        try:
            cdata = _cc_signature_normalize(_cc_deserialize_compact(sig))[1]
            der = _cc_cdata_to_der(cdata)
            # 0x04 == uncompressed-point prefix; ecdsa's to_string() omits it.
            return _CCPublicKey(b"\x04" + pub).verify(der, message, hasher=_sha256d)
        except Exception:
            return False

    try:
        vk = VerifyingKey.from_string(pub, curve=SECP256k1)
        return vk.verify(sig, message, hashfunc=hashlib.sha256)
    except (BadSignatureError, ValueError, TypeError):
        return False


class Wallet:
    """A keypair and the address it controls. Generate a fresh one, or reload one
    from a stored private key. The private key never leaves the wallet except as
    an explicit export the caller asks for."""

    def __init__(self, signing_key: SigningKey):
        self._signing_key = signing_key
        self._verifying_key = signing_key.get_verifying_key()

    @classmethod
    def generate(cls) -> "Wallet":
        return cls(SigningKey.generate(curve=SECP256k1))

    @classmethod
    def from_private_key_hex(cls, private_key_hex: str) -> "Wallet":
        return cls(SigningKey.from_string(bytes.fromhex(private_key_hex), curve=SECP256k1))

    @property
    def private_key_hex(self) -> str:
        return self._signing_key.to_string().hex()

    @property
    def public_key_hex(self) -> str:
        return self._verifying_key.to_string().hex()

    @property
    def address(self) -> str:
        return address_from_public_key(self._verifying_key.to_string())

    def sign(self, message: bytes) -> str:
        """Sign the canonical bytes of a transaction (see Transaction.signing_bytes)
        and return the signature as hex. Deterministic (RFC 6979) over SHA-256 —
        strong (the library's bare sign() defaults to SHA-1) and reproducible, so
        an independent implementation (e.g. the browser wallet) that follows the
        same rule produces a byte-identical signature."""
        return self._signing_key.sign_deterministic(message, hashfunc=hashlib.sha256).hex()

File: frontend/src/services/wallet.js
// DeCoin browser wallet — real ECDSA over secp256k1, matching src/wallet.py.
//
// The node enforces signatures, so the frontend must sign transactions with a
// real key, not invent an address as 'DEC' + Math.random(). This produces the
// SAME addresses, canonical bytes, and (deterministic) signatures the Python
// node verifies — cross-checked byte-for-byte in frontend/test/wallet.test.mjs.
import { secp256k1 } from '@noble/curves/secp256k1.js'
import { sha256 } from '@noble/hashes/sha2.js'
import { ripemd160 } from '@noble/hashes/legacy.js'
import { bytesToHex, hexToBytes, utf8ToBytes } from '@noble/hashes/utils.js'

const ADDRESS_PREFIX = 'DEC'

// Public keys are the 64-byte x||y form (no 0x04 prefix), matching Python's
// VerifyingKey.to_string().
export function addressFromPublicKeyHex(publicKeyHex) {
  const pub = hexToBytes(publicKeyHex)
  return ADDRESS_PREFIX + bytesToHex(ripemd160(sha256(pub)))
}

// Canonical JSON identical to Python's json.dumps(obj, sort_keys=True):
// keys sorted (recursively), ", " between items and ": " between key and value.
// Data is ASCII (addresses, integer amounts) so escaping matches on both sides.
export function canonicalJson(value) {
  if (value === null) return 'null'
  const t = typeof value
  if (t === 'number') {
    if (!Number.isInteger(value)) throw new Error('canonicalJson: only integers are allowed')
    return String(value)
  }
  if (t === 'boolean') return value ? 'true' : 'false'
  if (t === 'string') return JSON.stringify(value)
  if (Array.isArray(value)) return '[' + value.map(canonicalJson).join(', ') + ']'
  const keys = Object.keys(value).sort()
  return '{' + keys.map((k) => JSON.stringify(k) + ': ' + canonicalJson(value[k])).join(', ') + '}'
}

// Metadata keys that carry signatures rather than signed content (multisig).
const UNSIGNED_META_KEYS = new Set(['signatures', 'public_keys'])

// The exact bytes a signature commits to — mirrors Transaction.signing_bytes.
// The timestamp is signed as an INTEGER (whole seconds) so the encoding matches
// Python's (a float would serialize as "170...0.0" here vs "170...0" in JS).
export function signingBytes(tx) {
  const meta = {}
  for (const [k, v] of Object.entries(tx.metadata || {})) {
    if (!UNSIGNED_META_KEYS.has(k)) meta[k] = v
  }
  const data = {
    type: tx.type,
    sender: tx.sender,
    recipient: tx.recipient,
    amount: tx.amount,
    timestamp: Math.trunc(tx.timestamp),
    nonce: tx.nonce ?? 0,
    metadata: meta,
  }
  return utf8ToBytes(canonicalJson(data))
}

// Verify a signature the way the node does (prehash SHA-256). The stored public
// key is 64-byte x||y; noble wants the 0x04-prefixed uncompressed form.
export function verifySignature(publicKeyHex, signatureHex, messageBytes) {
  try {
    const pub = new Uint8Array([4, ...hexToBytes(publicKeyHex)])
    // lowS:false accepts non-normalised signatures, matching python-ecdsa.
    return secp256k1.verify(hexToBytes(signatureHex), messageBytes, pub, { prehash: true, lowS: false })
  } catch {
    return false
  }
}

// Build a signed standard transaction as the object POST /transaction expects.
// Amount, fee, and nonce are integers; the timestamp is whole seconds.
export function createSignedTransaction(wallet, { recipient, amount, nonce = 0, fee = 0, note } = {}) {
  const metadata = { fee }
  if (note) metadata.note = note
  const tx = {
    type: 'standard', // chain type value, used inside signing_bytes
    sender: wallet.address,
    recipient,
    amount,
    timestamp: Math.floor(Date.now() / 1000),
    nonce,
    metadata,
  }
  wallet.signTransaction(tx)
  return {
    sender: tx.sender,
    recipient: tx.recipient,
    amount: tx.amount,
    transaction_type: 'standard', // schema type value
    metadata: tx.metadata,
    nonce: tx.nonce,
    timestamp: tx.timestamp,
    signature: tx.signature,
    public_key: tx.public_key,
  }
}


export class Wallet {
  constructor(privateKeyHex) {
    this._priv = hexToBytes(privateKeyHex)
  }

  static generate() {
    return new Wallet(bytesToHex(secp256k1.utils.randomSecretKey()))
  }

  static fromPrivateKeyHex(hex) {
    return new Wallet(hex)
  }

  get privateKeyHex() {
    return bytesToHex(this._priv)
  }

  get publicKeyHex() {
    // 0x04 || x || y, then drop the prefix byte to match Python's 64-byte form.
    return bytesToHex(secp256k1.getPublicKey(this._priv, false).slice(1))
  }

  get address() {
    return addressFromPublicKeyHex(this.publicKeyHex)
  }

  // Deterministic (RFC 6979) signature over SHA-256, raw r||s hex. lowS:false so
  // it matches Python's ecdsa (which does not low-s normalise).
  sign(messageBytes) {
    // prehash:true lets noble hash the message with SHA-256 internally, which
    // reproduces python-ecdsa's sign_deterministic(hashfunc=sha256) byte for
    // byte (hashing first and passing the digest takes a different RFC 6979
    // path). lowS:false because python-ecdsa does not low-s normalise. Returns
    // the compact 64-byte r||s signature.
    return bytesToHex(secp256k1.sign(messageBytes, this._priv, { lowS: false, prehash: true }))
  }

  // Sign a plain transaction object in place, attaching public_key + signature.
  signTransaction(tx) {
    tx.public_key = this.publicKeyHex
    tx.signature = this.sign(signingBytes(tx))
    return tx
  }
}
